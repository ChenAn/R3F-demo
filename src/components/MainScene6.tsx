import React from "react";
import * as THREE from 'three';
import * as Fiber from '@react-three/fiber';
import * as Drei from '@react-three/drei';
import DragableSphere from "./DragableSphere";
import { OrbitControls } from "three-stdlib";

const MainScene6 = () => {
  const controlRef = React.useRef<OrbitControls>(null);
  
  const onBoxClicked = (event: Fiber.ThreeEvent<MouseEvent>) => {
    const targetMaterial = (event.object as THREE.Mesh).material as THREE.MeshBasicMaterial;
    const currentColor = targetMaterial.color.getHex();
    targetMaterial.color = new THREE.Color((currentColor << 8) | (currentColor >> (24 - 8)));
  };

  const onSphereDragStart = () => {
    if (controlRef.current !== null) controlRef.current.enabled = false;
  }

  const onSphereDragEnd = () => {
    if (controlRef.current !== null) controlRef.current.enabled = true;
  }

  const onSphereDrag = (event: Fiber.ThreeEvent<PointerEvent>) => {
    // moveイベントがCubeにも当たったかを確認
    const modelIntersection = event.intersections.find((intersection) => {
      return intersection.eventObject.name === 'bigCube';
    });
    // 球体を当たったCubeの表面位置に移動
    if (modelIntersection) {
      event.eventObject.position.copy(modelIntersection.point);
    }
  };

  // これは削除したら↑のイベントからbigCubeが出てこない
  const onPointerMove = () => {
  };

  return (
    <>
      <Drei.PerspectiveCamera makeDefault fov={60} position={[0, 0, 10]}/>
      <Drei.Box args={[2, 2, 2]} onClick={onBoxClicked} name='bigCube' 
        onPointerMove={onPointerMove}>
        <meshBasicMaterial color="red" />
      </Drei.Box>
      <DragableSphere position={[0, 0, 1]} radius={0.25} onDrag={onSphereDrag}
        onDragStart={onSphereDragStart} onDragEnd={onSphereDragEnd}/>
      <Drei.OrbitControls ref={controlRef} makeDefault minDistance={2.0} maxDistance={50.0}/>
    </>
  );
};

export default MainScene6;
