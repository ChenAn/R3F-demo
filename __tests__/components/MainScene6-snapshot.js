import React from 'react';
import ReactThreeTestRenderer from '@react-three/test-renderer';
import MainScene6 from '@src/components/MainScene6';

// eslint-disable-next-line import/no-extraneous-dependencies
jest.mock('scheduler', () => require('scheduler/unstable_mock'));

describe('MainScene6', () => {
  it('初期表示', async () => {
    const renderer = await ReactThreeTestRenderer.create(
      <MainScene6/>,
    );
    expect(renderer.toTree()).toMatchSnapshot();
  });
});