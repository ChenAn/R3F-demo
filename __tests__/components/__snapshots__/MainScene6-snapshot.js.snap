// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScene6 初期表示 1`] = `
Array [
  Object {
    "children": Array [],
    "props": Object {
      "fov": 60,
      "position": Array [
        0,
        0,
        10,
      ],
    },
    "type": "perspectiveCamera",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "props": Object {
          "args": Array [
            2,
            2,
            2,
          ],
          "attach": "geometry",
        },
        "type": "boxGeometry",
      },
      Object {
        "children": Array [],
        "props": Object {
          "attach": "material",
          "color": "red",
        },
        "type": "meshBasicMaterial",
      },
    ],
    "props": Object {
      "name": "bigCube",
      "onClick": [Function],
      "onPointerMove": [Function],
    },
    "type": "mesh",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "props": Object {
          "args": Array [
            0.25,
            32,
            32,
          ],
          "attach": "geometry",
        },
        "type": "sphereGeometry",
      },
      Object {
        "children": Array [],
        "props": Object {
          "attach": "material",
          "color": "white",
        },
        "type": "meshBasicMaterial",
      },
    ],
    "props": Object {
      "onClick": [Function],
      "onPointerDown": [Function],
      "onPointerMove": [Function],
      "onPointerUp": [Function],
      "position": Array [
        0,
        0,
        1,
      ],
    },
    "type": "mesh",
  },
]
`;
